infra/backup_tools/backup.sh

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env bash
+#
+# Copyright 2019 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# USAGE NOTES
+#
+# Backs up prod registries. This is a thin orchestrator as all the heavy lifting
+# is done by copy_with_date.sh and record_consistency.sh. The only required
+# argument is the path to the service account that has write access to the
+# backup GCRs.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+copy_with_date()
+{
+    local timestamp
+    local source_gcr_repo
+    local backup_gcr_repo
+    local sa_key_path
+
+    # Configure variables.
+    # We use a timestamp of the form YYYY/MM/DD/HH because this makes the backup
+    # folders more easily traversable from a human perspective.
+    timestamp="$(date -u +"%Y/%m/%d/%H")"
+    source_gcr_repo="${1}" # "us.gcr.io/k8s-artifacts-prod"
+    backup_gcr_repo="${2}" # "us.gcr.io/k8s-artifacts-prod-backup"
+    sa_key_path="${3}"
+
+    # Perform backup by copying all images recursively over.
+    docker run \
+        -v "${sa_key_path}":/auth.json \
+        --env GOOGLE_APPLICATION_CREDENTIALS=/auth.json \
+        gcr.io/go-containerregistry/gcrane \
+            cp -r "${source_gcr_repo}" "${backup_gcr_repo}/${timestamp}"
+}
+
+if (( $# != 1 )); then
+    "usage: ./backup.sh <svc_acct_key_path_for_backup_gcr>"
+    exit 1
+fi
+
+prod_repos=(
+    asia.gcr.io/k8s-artifacts-prod
+    eu.gcr.io/k8s-artifacts-prod
+    us.gcr.io/k8s-artifacts-prod
+)
+sa_key_path="${1}"
+
+# Activate creds.
+gcloud auth activate-service-account --key-file "${sa_key_path}"
+
+# Copy each region to its backup.
+for prod_repo in "${prod_repos[@]}"; do
+    copy_with_date "${prod_repo}" "${prod_repo}-backup" "${sa_key_path}"
```

Comment:
NB I changed "-backup" to "-bak" in latest push of the ensure script (because we overflowed allowable project names :)

Reply:
ACK